### DISCUSSION TITLE
[Modification] - Allow users to modify the rendered message for input instructions

### DISCUSSION THREAD

**Discussion by HYP3R00T** [petereon/beaupy#106] [Ideas]
I would very much like the option to customize the colours of the instruction we get at the end of `select` and `select_multiple`. Bold with White colour is something I don't want because it looks highlighted. A muted colour would be better. I know it's challenging to keep the whole codebase unopinionated. But modifying the style is something I want. 

If you allow, I would like to work on this. But before I get started and go for a pull request, I want to know your opinion.

Things to modify:
- Custom instructions 
- Custom encapsulation string for `select_multiple`. (We have square brackets inside which we can add any custom text, but I want to remove the square bracket or turn it into red just for me).

**Comment by petereon** [comment]
I definitely see your point. I am personally very in favour of customizability, BUT I am a little wary about including more of it to `beaupy` - I am of the opinion I stretched it a little too much already - as I want to target beginners to the world of programming and CLIs having their start in making CLI apps as simple with as little fiddling as possible.

That disclaimer of the way, a little self-advertisement if you allow: I wrote another Python library dedicated to CLIs - [questo](https://github.com/petereon/questo) - that's used as an underlying engine of `beaupy` since some time ago. It's very simple in its basic design. You have a CLI element which has some sort of a state. This state can be modified by a _keypress handler_ (you can define your own and there is a sane default available for both `select` and `prompt`). Now best of all, you can define your own _renderer_ that takes the state and converts it to whatever textual representation will be displayed in the terminal (again sane defaults are available). You can use the whole power of [`rich`](https://github.com/Textualize/rich) library to make these textual representations as simple or as complex as you wish. I believe this might already have your use-case covered. Only drawback being that it's not documented too well, but you can take a look at the example in `questo` README and the source code of `beaupy` for reference.

With that, if you are still set on modifying `beaupy` to support your use-case, I am on board with the following:
- Adding parameter `hint_text` to each element that controls what hint is displayed and defaults to whatever is there currently
- Adding flag parameter `display_hint` to each element that has a hint which default to `True`
- Adding flag parameter (not sure about a good name here) which decides whether the square brackets in `select_multiple` are displayed

**Reply by Samuel-1009** [reply]
Exactly the same issue here. I also want the instructions at the end of `select` and `select_multiple` to be customizable. I understand your opinion, but I think that people would still favor customizability more. Also, I want the formatting in the options of the `select` function to be more customizable, with the help of: 

- Parameter `selected_option` which will by default be `"> {}"` where the braces can be replaced with the text of the selected option, similar to the way in which the `format` function replaces the braces with the desired value. With the help of this, we can add cursors on both the sides, and change colors of the text and the cursor according to our wish.
- Parameter `deselected_option` which will work similarly, but will control the formattings of the options which are not selected.

I think I added way too much complexity already, I don't even think that it's necessary. The main thing that I want is the instructions at the end of `select` and `select_multiple` to be customizable, but if you still think that it adds more complexity to the code, then its better if it isn't implemented.

**Reply by petereon** [reply]
Hey @Samuel-1009, thank you for reaching out and a feature request. I am inclined to implement the customizable instruction as you suggest. Progress will be tracked in the following issue I have just created: https://github.com/petereon/beaupy/issues/135

As for your other 2 requests - these are rather specific cosmetics and while I understand that TUIs tend to be very visually opinionated, `beaupy` aims to be approachable and productive, providing sane details without having too much nitty-gritty to get tangled in.

If you look for customizability, [`questo`](https://github.com/petereon/questo) is as customizable as it gets. I think it separates functionality and representation fairly well, so you can just start from defaults and tinker with the default "renderer" to get it to your exact liking. It is poorly documented (I should seriously step my documentation game up), but the `beaupy` implementation is a reasonable reference.

EDIT: Just had a look ([link](https://github.com/petereon/beaupy/blob/e79cf248399e33da94bcfeb75bf2b70d74cec8d0/beaupy/_beaupy.py#L299)). It's a little hardcore to be fair 😅

**Comment by HYP3R00T** [comment]
As you planned to target beginners, I will respect that and won't be adding complexity to it. I plan to use `questo` and create the CLI tool that I need. I just realized `beaupy` is missing some features (for advanced users). 

> - Adding parameter `hint_text` to each element that controls what hint is displayed and defaults to whatever is there currently
> - Adding flag parameter `display_hint` to each element that has a hint which defaults to `True`

If you want, I can work on it. Otherwise, I can let it go. I have already created a tool, called [`hypercli`](https://github.com/HYP3R00T/hypercli) which doesn't support multiple selections for now. It allows the users to navigate back and forth though. The project I am working on now depends a lot on the `select_multiple` feature of `beaupy`. 

Missing features in `beaupy`:
- Styling each element
- Banners like the name of the project at the top with pyfiglet texts.
- Navigation to the previous menu
- Single key press to terminate the program

If you ever want to work on any of these missing features, please tag me. I would love to work on this.
